lib/widgets/message_list.dart

```diff
@@ -725,12 +726,47 @@ class RecipientHeaderDate extends StatelessWidget {
           //   https://developer.mozilla.org/en-US/docs/Web/CSS/font-variant-caps#all-small-caps
           fontFeatures: const [FontFeature.enable('c2sc'), FontFeature.enable('smcp')],
         ).merge(weightVariableTextStyle(context)),
-        _kRecipientHeaderDateFormat.format(
-          DateTime.fromMillisecondsSinceEpoch(message.timestamp * 1000))));
+        formatHeaderDate(
+          zulipLocalizations,
+          DateTime.fromMillisecondsSinceEpoch(message.timestamp * 1000),
+          now: DateTime.now())));
   }
 }
 
-final _kRecipientHeaderDateFormat = DateFormat('y-MM-dd', 'en_US'); // TODO(#278)
+@visibleForTesting
+String formatHeaderDate(
+  ZulipLocalizations zulipLocalizations,
+  DateTime dateTime, {
+  required DateTime now,
+}) {
+  assert(!dateTime.isUtc && !dateTime.isUtc,
+    '`dateTime` and `now` need to be in local time.');
```

Comment:
This worried me for a bit — what if local time happens to be UTC?

But it looks like it's fine: `isUtc` is true only if the explicit `isUtc` was set to true at construction. Here's a test program:
```dart
void main() {
  final t1 = DateTime.fromMillisecondsSinceEpoch(1701739446000);
  final t2 = DateTime.fromMillisecondsSinceEpoch(1701739446000, isUtc: true);
  print("$t1 timeZoneName: ${t1.timeZoneName} isUtc: ${t1.isUtc}");
  print("$t2 timeZoneName: ${t2.timeZoneName} isUtc: ${t2.isUtc}");
}
```

And here's output:
```
$ TZ=UTC dart /tmp/time.dart 
2023-12-05 01:24:06.000 timeZoneName: UTC isUtc: false
2023-12-05 01:24:06.000Z timeZoneName: UTC isUtc: true
```